File: server/internal/expenses/expense_item.go
package expenses

import (
	"math/big"
	"time"

	"gorm.io/gorm"
)

type ExpenseItem struct {
	ID            uint           `json:"id" gorm:"primaryKey"`
	BillPaymentID *uint          `json:"bill_payment_id" gorm:"index"` // optional, can be null
	ExpenseTypeID uint           `json:"expense_type_id" gorm:"not null;index"`
	Year          int            `json:"year" gorm:"not null"`                      // format: YYYY
	Month         int            `json:"month" gorm:"not null"`                     // format: MM
	Amount        string         `json:"amount" gorm:"type:decimal(10,2);not null"` // Store as string for precision
	Note          string         `json:"note"`
	UserID        uint           `json:"user_id" gorm:"not null;index"`
	CreatedAt     time.Time      `json:"created_at"`
	UpdatedAt     time.Time      `json:"updated_at"`
	DeletedAt     gorm.DeletedAt `json:"-" gorm:"index"`

	// Associations
	ExpenseType ExpenseType `json:"expense_type,omitempty" gorm:"foreignKey:ExpenseTypeID"`
	BillPayment BillPayment `json:"bill_payment,omitempty" gorm:"foreignKey:BillPaymentID"`
}

// GetAmountAsBigFloat converts the stored string amount to big.Float
func (ei *ExpenseItem) GetAmountAsBigFloat() *big.Float {
	amount := new(big.Float)
	amount.SetString(ei.Amount)
	return amount
}

// SetAmountFromBigFloat sets the amount from a big.Float
func (ei *ExpenseItem) SetAmountFromBigFloat(amount *big.Float) {
	ei.Amount = amount.String()
}

File: ios/jiceotapp/Models/ExpenseItem.swift
//
//  ExpenseItem.swift
//  JiceotApp
//
//  Created on 12/11/2025.
//

import Foundation

struct ExpenseItem: Codable, Identifiable {
    let id: Int
    let billPaymentId: Int?
    let billTypeId: Int?
    let expenseTypeId: Int
    let year: Int
    let month: Int
    let amount: String
    let note: String
    let userId: Int
    let createdAt: String
    let updatedAt: String
    let expenseType: ExpenseType?
    let billPayment: BillPayment?
    let billType: BillType?
    
    enum CodingKeys: String, CodingKey {
        case id
        case billPaymentId = "bill_payment_id"
        case billTypeId = "bill_type_id"
        case expenseTypeId = "expense_type_id"
        case year
        case month
        case amount
        case note
        case userId = "user_id"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case expenseType = "expense_type"
        case billPayment = "bill_payment"
        case billType = "bill_type"
    }
}

struct CreateExpenseItemRequest: Codable {
    let billPaymentId: Int?
    let billTypeId: Int?
    let expenseTypeId: Int
    let year: Int
    let month: Int
    let amount: String
    let note: String?
    
    enum CodingKeys: String, CodingKey {
        case billPaymentId = "bill_payment_id"
        case billTypeId = "bill_type_id"
        case expenseTypeId = "expense_type_id"
        case year
        case month
        case amount
        case note
    }
}

struct ExpenseItemListResponse: Codable {
    let expenseItems: [ExpenseItem]
    let total: Int
    
    enum CodingKeys: String, CodingKey {
        case expenseItems = "expense_items"
        case total
    }
}
